Armadillo/src/main/java/com/scribd/armadillo/playback/PlaybackEngine.kt

```diff
@@ -111,18 +124,22 @@ internal class ExoplayerPlaybackEngine(private var audioPlayable: AudioPlayable)
 
     override fun beginPlayback(isAutoPlay: Boolean, maxDurationDiscrepancy: Int, initialOffset: Milliseconds) {
         stateModifier.dispatch(NewAudioPlayableAction(audioPlayable, maxDurationDiscrepancy, initialOffset))
-
+        clearExistingExoPlayers()
         exoPlayer = createExoplayerInstance(context, audioAttributes.exoPlayerAttrs)
 
         val mediaSource = MediaSourceRetrieverImpl()
             .generateMediaSource(audioPlayable.request, context)
         exoPlayer.setMediaSource(mediaSource)
         exoPlayer.prepare()
+        playerEventListener = PlayerEventListener()
+        exoplayerReferences.add(ExoPlayerReference(exoPlayer, playerEventListener))
```

Comment:
Allowing multiple instances of exoplayer seems like a big red flag

Reply:
It also seems like you're never actually going to have more than one in the list, because you always clear it and then make a new one

Reply:
So, if the `PlaybackService` is legitimately killed via [this post](https://stackoverflow.com/questions/43410175/how-to-stop-mediabrowserservicecompat), the exoplayer instance is not `deinit`'d, which means it continues running and being active even after playing a new audiobook. So multiple exoplayers could still be running in that case. Note that `onDestroy()` in services is not reliably called.  I think this is a much less common instance than what we see in firebase.

Killing the service in this way does disable playback controls.

However, I've found that I can replicate the original crash/error by repeatedly calling `stopForeground(removeNotification = true)` while the app is in the background. When the crash finally occurs, and audio still continues, killing the app and re-entering it returns the progress state to the moment that the the crash occurred. However, doing subsequent checks to see if this service is still running returns `true`. 

"Killing" the service in this way does not disable playback controls.

I was able to find that the `ForegroundServiceStartNotAllowedException` throwing an exception in `PlaybackActionTransmitter` was killing the armadilloState polling. By adding a try/catching `ForegroundServiceStartNotAllowedException` and letting all other errors through, progress is preserved even after  the exception. This seems to be a more elegant fix, but still doesn't solve the exception in the first place. I'm going to look into where we might be unnecessarily/overly calling `stopForeground(true)` to remedy this, but we should patch up this progress issue in the meantime.